pkg/dns/aws/dns.go

```diff
@@ -242,83 +242,101 @@ func (m *Provider) change(record *iov1.DNSRecord, zone configv1.DNSZone, action
+	m.lock.Lock()
+	updatedRecords := sets.NewString(m.updatedRecords.List()...)
+	m.lock.Unlock()
 	if len(domain) == 0 {
 		return fmt.Errorf("domain is required")
 	}
-	if len(target) == 0 {
+	if len(targets) == 0 {
 		return fmt.Errorf("target is required")
 	}
 
+	var changes []*route53.Change
+
 	zoneID, err := m.getZoneID(zone)
 	if err != nil {
 		return fmt.Errorf("failed to find hosted zone for record: %v", err)
 	}
 
-	// Find the target hosted zone of the load balancer attached to the service.
-	targetHostedZoneID, err := m.getLBHostedZone(target)
-	if err != nil {
-		return fmt.Errorf("failed to get hosted zone for load balancer target %q: %v", target, err)
-	}
+	for _, target := range targets {
+		var key string
+		// Find the target hosted zone of the load balancer attached to the service.
+		targetHostedZoneID, err := m.getLBHostedZone(target)
+		if err != nil {
+			return fmt.Errorf("failed to get hosted zone for load balancer target %q: %v", target, err)
+		}
+		key = zoneID + domain + target
+		// Configure records and cache updates.
+		// TODO: handle the caching/diff detection in a better way.
+		m.lock.Lock()
+		alreadyHaveRecord := updatedRecords.Has(key)
+		m.lock.Unlock()
+		// Only process updates once for now because we're not diffing.
+		if alreadyHaveRecord && action == upsertAction {
+			log.V(2).Info("skipping cached DNS record update", "record", record)
+			continue
+		}
 
-	// Configure records and cache updates.
-	// TODO: handle the caching/diff detection in a better way.
-	m.lock.Lock()
-	defer m.lock.Unlock()
-	key := zoneID + domain + target
-	// Only process updates once for now because we're not diffing.
-	if m.updatedRecords.Has(key) && action == upsertAction {
-		log.V(2).Info("skipping cached DNS record update", "record", record)
-		return nil
-	}
-	err = m.updateAlias(domain, zoneID, target, targetHostedZoneID, string(action))
-	if err != nil {
-		return fmt.Errorf("failed to update alias in zone %s: %v", zoneID, err)
+		switch action {
+		case upsertAction:
+			updatedRecords.Insert(key)
+			log.Info("upserted DNS record", "record", record)
+		case deleteAction:
+			updatedRecords.Delete(key)
+			log.Info("deleted DNS record", "record", record)
+		}
+
+		change := &route53.Change{
+			Action: aws.String(string(action)),
+			ResourceRecordSet: &route53.ResourceRecordSet{
+				Name: aws.String(domain),
+				Type: aws.String("A"),
+				AliasTarget: &route53.AliasTarget{
+					HostedZoneId:         aws.String(targetHostedZoneID),
+					DNSName:              aws.String(target),
+					EvaluateTargetHealth: aws.Bool(false),
+				},
+			},
+		}
+
+		changes = append(changes, change)
 	}
-	switch action {
-	case upsertAction:
-		m.updatedRecords.Insert(key)
-		log.Info("upserted DNS record", "record", record)
-	case deleteAction:
-		m.updatedRecords.Delete(key)
-		log.Info("deleted DNS record", "record", record)
+
+	if len(changes) > 0 {
+		m.lock.Lock()
+		m.updatedRecords = updatedRecords
+		m.lock.Unlock()
```

Comment:
Don't update `m.updatedRecords` till after calling `m.updateAlias` in case the update fails.

Reply:
done